torch/nn/modules/module.py

```diff
@@ -382,6 +382,141 @@ def add_module(self, name: str, module: Optional['Module']) -> None:
             raise KeyError("module name can't be empty string \"\"")
         self._modules[name] = module
 
+    def has_submodule(self, target: str) -> bool:
```

Comment:
@jbschlosser can you take a look at these changes? We added these methods to `fx.GraphModule`, but they only manipulate the module hierarchy, so it seemed like they could be equally applicable to `nn.Module` generally. Do these look OK?

Reply:
Are these methods being added to support fx-related stuff or for utility in the public API?

If the latter, I'm a bit ambivalent about how they fit with the traditional usage of `Module`s. In general, the recommendation is to define custom modules deriving from the base `torch.nn.Module` that implement the `forward()` function for their computation.

Since the `forward()` function can't be "rewritten", submodules cannot in general be added / removed dynamically without invalidating the `forward()` logic (at least for modules that aren't `nn.Sequential`-like containers AKA ones that don't simply iterate through submodules in their `forward()` implementation).

For example, I'm worried that someone will think it's perfectly reasonable to delete the final classification layer of their [ResNet module](https://github.com/pytorch/vision/blob/2e8c124ff7266437efd37c7a23684f64fe6904cd/torchvision/models/resnet.py#L244) to try to obtain image embeddings and then get confused when an error occurs during the forward pass.

Will discuss this more with the other nn folks tomorrow!

Reply:
I would agree with @jbschlosser  here that these API look very surprising for nn.Module.
Another thing I am weary about as well is that they allow users to create invalid nn.Modules very easily. Which would hint to the fact that they should not be public.

Also, for the namings issue with nesting, `.named_modules()` should give you all the info you need. In particular you can check the keys to know if a submodule exist in a nested way.

Reply:
@albanD `.named_modules()` has performance issues in that a query against it is `O(N)` in the number of transitive modules. We're already seeing poor performance in internal ads workloads because of it. I agree that the submodule manipulation APIs here don't make sense, but it's pretty critical that we have a way to query for the existence of a submodule without blowing up the complexity of the program

Reply:
That said, I'm okay with moving all of this back to `GraphModule` to unblock, but I thought I'd make you aware of the issue

Reply:
Yes thanks for sharing!
As Joel mentioned above, the getter sounds like it would be useful just for that reason so that one would be ok to add I think.
But the setter looks much more dangerous and out of place.